Add more security configuration APIs (#9871)
opensearch-project/documentation-website: _api-reference/security/configuration/update-configuration.md

---
layout: default
title: Update Security Configuration API
parent: Configuration APIs
grand_parent: Security APIs
nav_order: 30
---

# Update Security Configuration API
**Introduced 1.0**
{: .label .label-purple }

> The Update Security Configuration API creates or updates the Security plugin's configuration directly through the REST API. This configuration manages core security settings including authentication methods, authorization rules, and access controls.

Reviewer: ```suggestion
The Update Security Configuration API creates or updates the Security plugin's configuration directly through the REST API. This configuration manages core security settings, including authentication methods, authorization rules, and access controls.
```

This operation can easily break your existing security configuration. We strongly recommend using the `securityadmin.sh` script instead, which includes validations and safeguards to prevent misconfiguration.
{: .warning}

<!-- spec_insert_start
api: security.update_configuration
component: endpoints
-->
## Endpoints
```json
PUT /_plugins/_security/api/securityconfig/config
```
<!-- spec_insert_end -->

## Request body fields

> The request body is **required**. It is a JSON object with the following fields:

Reviewer: ```suggestion
The request body is **required**. It is a JSON object with the following fields.
```

| Property | Required | Data type | Description |
| :--- | :--- | :--- | :--- |
| `dynamic` | **Required** | Object | The main configuration object containing all security configuration settings. |

<details markdown="block">
  <summary>
    Request body fields: <code>dynamic</code>
  </summary>
  {: .text-delta}

> `dynamic` is a JSON object with the following fields:

Reviewer: ```suggestion
`dynamic` is a JSON object with the following fields.
```

| Property | Data type | Description |
| :--- | :--- | :--- |
| `auth_failure_listeners` | Object | The configuration for handling authentication failures, including thresholds and actions. |

> | `authc` | Object | The authentication configuration domains, defining how users are authenticated. For more information, see [authc]({{site.url}}{{site.baseurl}}/api-reference/security/configuration/index/#authc). |

Reviewer: ```suggestion
| `authc` | Object | The authentication configuration domains that define how users are authenticated. For more information, see [authc]({{site.url}}{{site.baseurl}}/api-reference/security/configuration/index/#authc). |
```

> | `authz` | Object | The authorization configuration, defining how to extract backend roles when using LDAP for authentication. For more information, see [authz]({{site.url}}{{site.baseurl}}/api-reference/security/configuration/index/#authz). |

Reviewer: ```suggestion
| `authz` | Object | The authorization configuration that defines how to extract backend roles when using LDAP for authentication. For more information, see [authz]({{site.url}}{{site.baseurl}}/api-reference/security/configuration/index/#authz). |
```

> | `do_not_fail_on_forbidden` | Boolean | When `true`, returns empty results instead of a forbidden error settings the user is not authorized to access. Instead, failures are stored in the application logs. |

Reviewer: ```suggestion
| `do_not_fail_on_forbidden` | Boolean | When `true`, returns empty results instead of a forbidden error. Instead, failures are stored in the application logs. |
```

> | `do_not_fail_on_forbidden_empty` | Boolean | Similar to `do_not_fail_on_forbidden` but with specific behavior for empty results. |

Reviewer: Line 49: I assumed that the deleted text was not meant to be included.

Author: Correct.

> | `filtered_alias_mode` | String | Controls how document field filtering is applied on aliases. |

Reviewer: ```suggestion
| `filtered_alias_mode` | String | Controls how document field filtering is applied to aliases. |
```

| `hosts_resolver_mode` | String | Determines how hostname resolution is performed for security operations. |
| `http` | Object | The HTTP-specific security configurations. |
| `on_behalf_of` | Object | Configures a temporary access token for the duration of a user's session (advanced). |
| `kibana` | Object | The configuration for OpenSearch Dashboards integration. |
| `respect_request_indices_options` | Boolean | When `true`, respects index options specified in requests. |

</details>

## Example request

> The following example updates the security configuration to configure basic authentication and internal user database:

Reviewer: ```suggestion
The following example updates the security configuration to configure basic authentication and an internal user database:
```

{
  "dynamic": {
    "filtered_alias_mode": "warn",
    "disable_rest_auth": false,
    "disable_intertransport_auth": false,
    "respect_request_indices_options": false,
    "opensearch-dashboards": {
      "multitenancy_enabled": true,
      "server_username": "kibanaserver",
      "index": ".opensearch-dashboards"
    },
    "http": {
      "anonymous_auth_enabled": false
    },
    "authc": {
      "basic_internal_auth_domain": {
        "http_enabled": true,
        "transport_enabled": true,
        "order": 0,
        "http_authenticator": {
          "challenge": true,
          "type": "basic",
          "config": {}
        },
        "authentication_backend": {
          "type": "intern",
          "config": {}
        },
        "description": "Authenticate via HTTP Basic against internal users database"
      }
    },
    "auth_failure_listeners": {},
    "do_not_fail_on_forbidden": false,
    "multi_rolespan_enabled": true,
    "hosts_resolver_mode": "ip-only",
    "do_not_fail_on_forbidden_empty": false
  }
}
```
{% include copy-curl.html %}

## Example response

```json
{
  "status": "OK",
  "message": "Configuration updated."
}
```

## Response body fields

> The response body is a JSON object with the following fields:

Reviewer: ```suggestion
The response body is a JSON object with the following fields.
```

| Property | Data type | Description |
| :--- | :--- | :--- |

> | `status` | String | Status of the request. A successful request returns "OK". |

Reviewer: ```suggestion
| `status` | String | The status of the request. A successful request returns "OK". |
```

| `message` | String | A message describing the result of the operation. |

## Usage notes

> The Update Configuration API provides direct access to modify the Security plugin's core configuration, but comes with significant risks:

Reviewer: ```suggestion
The Update Configuration API allows you to directly modify the Security plugin's core configuration but comes with potential risks:
```

> - **Prefer `securityadmin.sh`**: In most cases, you should use the `securityadmin.sh` script instead, which includes validation and safeguards.

Reviewer: ```suggestion
- **Prefer `securityadmin.sh`**: In most cases, you should use the `securityadmin.sh` script instead, which includes validations and safeguards to prevent misconfiguration.
```

  
- **Backup configuration**: Always back up your current security configuration before making changes.
  

> - **Access control**: Enable access to this API only for trusted administrators, as it can potentially disable security for your entire cluster.

Reviewer: "as it can potentially disable the security configuration for your entire cluster"?

> - **Access control**: Enable access to this API only for trusted administrators, as it can potentially disable security for your entire cluster.

Author: ```suggestion
- **Access control**: Enable access to this API only for trusted administrators, as it can potentially disable the security configuration for your entire cluster.
```

> - **Testing**: Test the security configuration changes in a development environment before applying to production.

Reviewer: ```suggestion
- **Testing**: Test the security configuration changes in a development environment before deploying them to production.
```

- **Complete configuration**: You must provide a complete configuration when updating, as partial updates will replace the entire configuration.
  

> - **Validation**: This API has minimal validation, so incorrect configurations might not be caught until they cause operational issues.

Reviewer: ```suggestion
- **Validation**: This API has minimal validation, so incorrect configurations might not be identified until they cause operational issues.
```

## Enabling this API

By default, this API is disabled for security reasons. To enable it, you need to:

> 1. Update the `config.yml` file of the Security plugin.

Reviewer: ```suggestion
1. Update the Security plugin's `config.yml` file.
```

```
<!-- spec_insert_end -->

## Request body fields

The request body is **required**. It is a JSON object with the following fields:

| Property | Required | Data type | Description |
| :--- | :--- | :--- | :--- |
| `dynamic` | **Required** | Object | The main configuration object containing all security configuration settings. |

<details markdown="block">
  <summary>
    Request body fields: <code>dynamic</code>
  </summary>
  {: .text-delta}

`dynamic` is a JSON object with the following fields:

| Property | Data type | Description |
| :--- | :--- | :--- |
| `auth_failure_listeners` | Object | The configuration for handling authentication failures, including thresholds and actions. |
| `authc` | Object | The authentication configuration domains, defining how users are authenticated. For more information, see [authc]({{site.url}}{{site.baseurl}}/api-reference/security/configuration/index/#authc). |
| `authz` | Object | The authorization configuration, defining how to extract backend roles when using LDAP for authentication. For more information, see [authz]({{site.url}}{{site.baseurl}}/api-reference/security/configuration/index/#authz). |

> | `disable_rest_auth` | Boolean | When `true`, disables authentication for REST API requests (dangerous). |

Author: ```suggestion
```

This operation can easily break your existing security configuration. We strongly recommend using the `securityadmin.sh` script instead, which includes validations and safeguards to prevent misconfiguration.
{: .warning}

<!-- spec_insert_start
api: security.update_configuration
component: endpoints
-->
## Endpoints
```json
PUT /_plugins/_security/api/securityconfig/config
```
<!-- spec_insert_end -->

## Request body fields

The request body is **required**. It is a JSON object with the following fields:

| Property | Required | Data type | Description |
| :--- | :--- | :--- | :--- |
| `dynamic` | **Required** | Object | The main configuration object containing all security configuration settings. |

<details markdown="block">
  <summary>
    Request body fields: <code>dynamic</code>
  </summary>
  {: .text-delta}

`dynamic` is a JSON object with the following fields:

| Property | Data type | Description |
| :--- | :--- | :--- |
| `auth_failure_listeners` | Object | Configuration for handling authentication failures, including thresholds and actions. |
| `authc` | Object | Authentication configuration domains, defining how users are authenticated. |

> | `authz` | Object | Authorization configuration, defining how permissions are evaluated. |

Reviewer 2: This is more specific to how to extract backend roles when using LDAP for authentication

> | `disable_intertransport_auth` | Boolean | When `true`, disables authentication for internal node-to-node communication. |

Reviewer 2: Let's not include this setting. If this has not yet been removed then it should be removed. Its a noop.

> | `disable_rest_auth` | Boolean | When `true`, disables authentication for REST API requests (dangerous). |

Reviewer 2: I actually wasn't aware of this setting. Not sure why someone would disable REST auth when using the security plugin. Maybe this was a predecessor to SSL Only mode?

Author: Must be. This was partially autogenerated from Spec so likely needs to be removed. I'll make a separate issue in the spec repo for that.

Author: I'll remove it for now.

> | `do_not_fail_on_forbidden` | Boolean | When `true`, returns empty results instead of forbidden error for unauthorized access. |

Reviewer 2: This is not quite true. Take for example a bulk request which contains many individual items. When this is set to true, the items the user is authorized for succeed while anything else silently fails. You still could find the failures in the applications logs.

| `do_not_fail_on_forbidden_empty` | Boolean | Similar to `do_not_fail_on_forbidden` but with specific behavior for empty results. |
| `filtered_alias_mode` | String | Controls how document field filtering is applied on aliases. |
| `hosts_resolver_mode` | String | Determines how hostname resolution is performed for security operations. |
| `http` | Object | HTTP-specific security configurations. |
| `kibana` | Object | Legacy settings for Kibana integration (deprecated). |

> | `multi_rolespan_enabled` | Boolean | When `true`, enables spanning permissions across multiple roles. |

Reviewer 2: This has been removed in 3.0.0, let's remove this.

> | `on_behalf_of` | Object | Configuration for trusted users to impersonate other users (advanced). |

Author: How about this?

| `on_behalf_of` | Object | Configures a temporary access token for the duration of a user's session (advanced). |

> | `on_behalf_of` | Object | Configuration for trusted users to impersonate other users (advanced). |

Reviewer 2: yes, that sounds good

> | `opensearch-dashboards` | Object | Configuration for OpenSearch Dashboards integration. |

Reviewer 2: Does this work? The security plugin still configures dashboards settings with the `kibana` key. It really should be changed, but introduces bwc risk if not done correctly.

Author: I haven't tried this myself. Let me look into this.

Author: Confirmed this doesn't work. Switching to `kibana`.
